File: csharp/Behavioral/19-State/State/IJobState.cs
namespace State;

// Declares every possible action. Each state implements only
// the ones valid from itself; the rest are rejected.
public interface IJobState
{
    string Name { get; }

    void Assign(JobCard job, string handyman);
    void Start(JobCard job);
    void Pause(JobCard job, string reason);
    void Resume(JobCard job);
    void Complete(JobCard job);
    void Approve(JobCard job);
}

File: typescript/Behavioral/19-State/State/IJobState.ts
import type { JobCard } from "../Context/JobCard";

// Declares every possible action. Each state implements only
// the ones valid from itself; the rest are rejected.
export interface IJobState {
  readonly name: string;

  assign(job: JobCard, handyman: string): void;
  start(job: JobCard): void;
  pause(job: JobCard, reason: string): void;
  resume(job: JobCard): void;
  complete(job: JobCard): void;
  approve(job: JobCard): void;
}
